### PR TITLE
Add default values in script

### PR DESCRIPTION
**Pull request by driazati**
Add support for default values on script functions and Modules

Followup to #11962

### CODE REVIEW

File: torch/csrc/jit/script/init.cpp [hunk 1]
```diff
@@ -363,6 +346,37 @@ Resolver pythonResolver(ResolutionCallback rcb) {
...
 }
 
+FunctionSchema getSchemaWithDefaults(
+    std::unordered_map<std::string, py::object>& default_args,
+    FunctionSchema& schema) {
+  std::vector<Argument> new_args;
+  for (auto& arg : schema.arguments) {
+    auto it = default_args.find(arg.name);
+    if (it != default_args.end()) {
+      IValue value = *constantToIValue(it->second);
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: There is no need for `constantToIValue`, we already have `inline IValue toIValue(py::handle obj, const TypePtr& type);` which is the correct call for this purpose. Most importantly, this pathway ensure that the returned `IValue` is in fact the right `type`. The refactored `constantToIValue` does not ensure that so this code can cause the program to throw an exception.

File: torch/csrc/jit/script/init.cpp [hunk 2]
```diff
@@ -363,6 +346,37 @@ Resolver pythonResolver(ResolutionCallback rcb) {
...
+      if (arg.type->isSubtypeOf(DynamicType::get()) && value.isScalar()) {
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: Why is this needed? We don't convert things to tensors by default in other cases, and in Python these numbers will be scalars not tensors on entry to the function.

File: torch/csrc/jit/script/init.cpp [hunk 3]
```diff
@@ -524,9 +538,10 @@ void initJitScriptBindings(PyObject* module) {
     .def("graph_for", [](Method& self, py::args args, py::kwargs kwargs) {
       return self.graph_for(createStackForSchema(self.getSchema(), std::move(args), std::move(kwargs)));
     })
-    .def("forward_schema", [](Method &self, Def &def, bool is_method) {
+    .def("forward_schema", [](Method &self, Def &def, py::object defaults, bool is_method) {
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: nit: `(Method &self, Def &def, const std::unordered_map<std::string, py::object>& defaults, bool is_method)`
enables you to remove the cast, and will improve the error message if this is called incorrectly.

File: test/test_jit.py [hunk 4]
```diff
@@ -1974,6 +1974,47 @@ def loop_use_test(y):
...
+        self.assertEqual(
+            bool_fn(torch.ones(1), torch.tensor(1), torch.tensor(True)),
+            torch.ones(1))
+
+    def test_default_values_module(self):
+        class Test(torch.jit.ScriptModule):
+            def __init__(self):
+                super(Test, self).__init__()
+
+            def forward(self, input, other=4):
```
Comment (zdevito) [inline test/test_jit.py]: There is no `@torch.jit.script_method` annotation here. I believe the pathway used by script method does not correctly capture the default arguments.

File: torch/jit/__init__.py [hunk 5]
```diff
@@ -643,7 +643,7 @@ def script(fn, optimize=True, _frames_up=0):
     # 2) Throwing everything away except for the graph 3) Creating a new
     # ScriptModule and dumping that graph in 4) Re-populating the schema
     # because it was lost doing the previous
-    mod.__getattr__('forward').forward_schema(ast, False)
+    mod.__getattr__('forward').forward_schema(ast, get_default_args(fn), False)
```
Comment (zdevito) [inline torch/jit/__init__.py]: Look for `_create_methods` in ScriptModule. Default arguments need to be added there as well.

**Review by zdevito — COMMENTED** [review]
Cool! This is almost ready to go -- I think there is some additional work to add support or @script_method defaults.

File: torch/csrc/jit/script/init.cpp (hunk 1 above)
Comment (driazati) [reply torch/csrc/jit/script/init.cpp]: Without type hints the schema I was getting for the functions in `test_default_values_function` looked like this
```
simple_fn(Tensor x, Tensor a=<default>, Tensor b=<default>, Tensor c=<default>) ->
bool_fn(Tensor x, Tensor a=<default>, Tensor flag=<default>) ->
```
Which would make `inline IValue toIValue(py::handle obj, const TypePtr& type)` fail since it would try to make an `autograd::Variable` from a scalar.

I added a check for the correct type + accompanying test for the correct types in the latest commit

File: torch/csrc/jit/pybind_utils.h [hunk 6]
```diff
@@ -54,6 +57,30 @@ inline void findErrorInKwargs(
 }
 } // namespace detail
 
+inline at::optional<IValue> constantToIValue(py::object obj) {
```
Comment (zdevito) [inline torch/csrc/jit/pybind_utils.h]: This function is now unneeded. Let's not move it into a header unnecessarily.

File: torch/csrc/jit/script/init.cpp [hunk 7]
```diff
@@ -373,6 +357,35 @@ Resolver pythonResolver(ResolutionCallback rcb) {
...
+  std::vector<Argument> new_args;
+  for (auto& arg : schema.arguments) {
+    auto it = default_args.find(arg.name);
+    if (it != default_args.end()) {
+      try {
+        IValue value = toIValue(it->second, arg.type);
+        new_args.push_back(
+            Argument(arg.name, arg.type, arg.N, value, arg.kwarg_only));
+      } catch (py::cast_error& e) {
+        throw ErrorReport()
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: The error report needs a SourceRange, otherwise the error message will not be associated with a location in the code. You can get the SourceRange from the corresponding argument in the Def object.

File: torch/csrc/jit/script/init.cpp [hunk 8]
```diff
@@ -534,9 +562,10 @@ void initJitScriptBindings(PyObject* module) {
     .def("graph_for", [](Method& self, py::args args, py::kwargs kwargs) {
       return self.graph_for(createStackForSchema(self.getSchema(), std::move(args), std::move(kwargs)));
     })
-    .def("forward_schema", [](Method &self, Def &def, bool is_method) {
-      auto schema = extractSchemaFromDef(def, is_method);
-      self.setSchema(schema);
+    .def("forward_schema", [](Method &self, Def &def, py::object defss, bool is_method) {
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: same nit as last review:
```
(Method &self, Def &def, FunctionDefaults defaults, bool is_method)
```
enables you to remove the cast, and will improve the error message if this is called incorrectly.

File: test/test_jit.py [hunk 9]
```diff
@@ -1982,6 +1982,69 @@ def python_op_name_test(y):
...
+                result = x + a
+            return result
+
+        self.assertExpectedGraph(bool_fn.graph, "bool")
+        self.assertEqual(bool_fn(torch.ones(1)), torch.ones(1) + 9)
+        self.assertEqual(
+            bool_fn(torch.ones(1), torch.tensor(1), torch.tensor(True)),
+            torch.ones(1))
+
+        if not PY2:
```
Comment (zdevito) [inline test/test_jit.py]: This doesn't work. Parsing of the file is not guarded by an if statement.
Use python 2 style type comments like seen in other examples:
```
def hints(x, a=0.5, b=10):  # noqa: E999
    # type: (Tensor, float, int) -> Tensor
    return x + a + b
```

File: torch/csrc/jit/script/init.cpp [hunk 10]
```diff
@@ -398,7 +411,11 @@ void initJitScriptBindings(PyObject* module) {
             auto self = has_self ? std::make_shared<ModuleValue>(m) : nullptr;
             return defineMethodsInModule(*m, script, pythonResolver(rcb), self);
           })
-      .def("_create_methods", [](std::shared_ptr<Module> m, const std::vector<Def>& defs, const std::vector<ResolutionCallback>& rcbs) {
+      .def("_create_methods", [](
+          std::shared_ptr<Module> m,
+          const std::vector<Def>& defs,
+          const std::vector<ResolutionCallback>& rcbs,
+          std::vector<FunctionDefaults> defaults) {
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: nit: `const std::vector<FunctionDefaults>& defaults`

File: torch/csrc/jit/script/init.cpp [hunk 11]
```diff
@@ -373,6 +357,35 @@ Resolver pythonResolver(ResolutionCallback rcb) {
   };
 }
 
+FunctionSchema getSchemaWithDefaults(
+    FunctionDefaults& default_args,
```
Comment (zdevito) [inline torch/csrc/jit/script/init.cpp]: nit: `const FunctionDefaults&`

**Review by zdevito — APPROVED** [review]
This looks pretty good. There are some minor nits, and  the python 2 tests need to pass.

**Review by apaszke — COMMENTED** [review]
A small note here is that if we really want to support any kind of default argument, then Python semantics for list defaults are pretty ridiculous. Try this:
```python
def f(x=[]):
    x.append(1)
    print(x)

for i in range(5):
    f()
```

**Comment by driazati** [comment]
It is a bit weird but we should follow Python's semantics here, right? As is right now the output is the same between script mode / Python.
